src/librustc_mir_build/build/mod.rs

```diff
@@ -21,13 +21,19 @@ use rustc_target::spec::PanicStrategy;
 
 use super::lints;
 
-crate fn mir_built(tcx: TyCtxt<'_>, def_id: LocalDefId) -> ty::steal::Steal<Body<'_>> {
-    tcx.alloc_steal_mir(mir_build(tcx, def_id))
+crate fn mir_built<'tcx>(tcx: TyCtxt<'tcx>, def: ty::WithOptConstParam<LocalDefId>) -> &'tcx ty::steal::Steal<Body<'tcx>> {
+    if def.const_param_did.is_none() {
+        if let const_param_did @ Some(_) = tcx.opt_const_param_of(def.did) {
```

Comment:
I just realized that this two-step process could probably be a helper method on `WithOptConstParam` (e.g. `if let Some(def) = def.upgrade_with_const_param(tcx)` or something like that, with the method describing this kind of use pattern).

Again, you can file this under future improvements.